### DISCUSSION TITLE
关于Rust调试的问题

### DISCUSSION THREAD

**Discussion by EricDecen** [z2o-k7e/halo2-step-by-step#55] [Q&A]
新学Rust和Halo2，且没有跟上学习进度，还未搞懂怎么调试。
还请各位前辈帮我指点一下，如何使用VScode调试Rust的入口。
目前VSCode中已经安装了C/C++（for Windows)、rust-analyzer等插件。
（1）如何调试？以及如何配置launch.json？
以对halo2-tutorials进行调试为例，按下F5后出现如下情况：
![image](https://github.com/zkp-co-learning/halo2-step-by-step/assets/148237745/135fa9c8-7fd0-43b2-aac5-b41f12ee5df9)
launch.json中的配置如下：
![image](https://github.com/zkp-co-learning/halo2-step-by-step/assets/148237745/dc599939-3824-4903-b52f-a03b8d8a2cb1)

（2）打印出table图形的test，如何显示出来？是否缺少了什么dependencies，还是需要做什么设置？
以chap1中的simple.rs为例
上面的test_chap_1可以被run test
![image](https://github.com/zkp-co-learning/halo2-step-by-step/assets/148237745/fb45726c-82be-4dd3-84b3-3d186d334c5e)
但后面的 plot_chap_1_circuit不能被run test
![image](https://github.com/zkp-co-learning/halo2-step-by-step/assets/148237745/0ed682ec-dc76-4c34-bef3-d5e6b327a294)

（3）一个rs文件，需要具体什么条件才能Run test，或是缺少了哪些操作的步骤？
用同一个rs（如simple.rs），copy一份后，这个新的rs却不能被调试。同 时也没有从插件而来的一些提示，这是为何？
![image](https://github.com/zkp-co-learning/halo2-step-by-step/assets/148237745/d91a5eda-bade-4ecb-8ef4-fdc1bd312380)

**Comment by EricDecen** [comment]
@Demian101 @dajuguan 关于Rust的新手(也是code新手）入门操作的问题，小白入门，问的有点粗浅，不好意思啦~

**Comment by Demian101** [comment]
(1) 
Debug mode 话，一般是根据 LLDB 和 .toml 文件自动生成的，我这边的配置 FYI 👍 
```
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "configurations": [
    {
      "type": "lldb",
      "request": "launch",
      "name": "Debug executable 'halo2_tutorials'",
      "cargo": {
        "args": [
          "build",
          "--bin=halo2_tutorials",
          "--package=halo2_tutorials"
        ],
        "filter": {
          "name": "halo2_tutorials",
          "kind": "bin"
        }
      },
      "args": [],
      "cwd": "${workspaceFolder}"
    },
    {
      "type": "lldb",
      "request": "launch",
      "name": "Debug unit tests in executable 'halo2_tutorials'",
      "cargo": {
        "args": [
          "test",
          "--no-run",
          "--bin=halo2_tutorials",
          "--package=halo2_tutorials"
        ],
        "filter": {
          "name": "halo2_tutorials",
          "kind": "bin"
        }
      },
      "args": [],
      "cwd": "${workspaceFolder}"
    }
  ]
}
```

-----


(2)
打印出 table 图形的test，如何显示出来？

首先，什么是 [cfg(test)] ？
- [cfg(test)] 是一个条件编译属性，意思是只有在执行 test 时，此模块代码才会被编译和执行
- 好处是，当你在普通的编译或生产环境下构建你的程序时，测试代码不会被包括进去，从而减少了编译时间和生成的可执行文件的大小。
- 注意到画图的 test 函数有一个 `cfg[feature=dev-graph]` 标识，说明这个函数只在 feature dev-graph 被启用时才编译。

这个 test 按钮应该是 rust-analyzer 的自动行为，类似封装了一层语法糖。可以不用管，因为换个编译器可能就没有了，

实际执行直接使用命令行：

```bash
cargo test test_chap_1
cargo test plot_chap_1_circuit --features dev-graph
```

**Reply by Demian101** [reply]
MacOS 、 LLDB

**Comment by Demian101** [comment]
（3）

观察你该目录的 mod.rs 文件，是否 mod 包含了你 new 的 file
